external-crates/move/documentation/book/src/control-flow/pattern-matching.md

```diff
@@ -0,0 +1,567 @@
+
+```move
+public enum MyEnum {
+    Variant(u64, bool),
+    OtherVariant(bool, u64),
+}
+
+public enum OtherEnum {
+    V(MyEnum)
+}
+
+public struct MyStruct {
+    x: u64,
+    y: u64,
+}
+
+// literal pattern
+1
+
+// constant pattern
+MyConstant
+
+// variable pattern
+x
+
+// wildcard pattern
+_
+
+// constructor pattern that matches `MyEnum::Variant` with the fields `1` and `true`
+MyEnum::Variant(1, true)
+
+// constructor pattern that matches `MyEnum::Variant` with the fields `1` and binds the second field's value to `x`
+MyEnum::Variant(1, x)
+
+// multi-arity wildcard pattern that matches multiple fields within the `MyEnum::Variant` variant
+MyEnum::Variant(..)
+
+// constructor pattern that matches the `x` field of `MyStruct` and binds the `y` field to `other_variable`
+MyStruct { x, y: other_variable }
+
+// at-pattern that matches `MyEnum::Variant` and binds the entire value to `x`
+x @ MyEnum::Variant(..)
+
+// or-pattern that matches either `MyEnum::Variant` or `MyEnum::OtherVariant`
+MyEnum::Variant(..) | MyEnum::OtherVariant(..)
+
+// Same as the above or-pattern, but with explicit wildcards
+MyEnum::Variant(_, _) | MyEnum::OtherVariant(_, _)
+
+// or-pattern that matches either `MyEnum::Variant` or `MyEnum::OtherVariant` and binds the u64 field to `x`
+MyEnum::Variant(x, _) | MyEnum::OtherVariant(_, x)
+
+// constructor pattern that matches `OtherEnum::V` and if the inner `MyEnum` is `MyEnum::Variant`
+OtherEnum::V(MyEnum::Variant(..))
+```
+
+More concisely we have the following grammar for patterns in Move:
+
+```bnf
+pattern = <literal>
+        | <constant>
+        | <variable>
+        | _
+        | C { inner-pattern, inner-pattern, ... } // where C is a struct or enum variant
+        | C ( inner-pattern, inner-pattern, ... ) // where C is a struct or enum variant
+        | C                                       // where C is an enum variant
+        | <variable> @ top-level-pattern
+        | pattern | pattern
+inner-pattern = pattern
+              | ..
+              | mut pattern
+```
+
+Patterns that contain variables bind them to the value being matched. These variables can then be
+used either in any match guard expressions, or on the right-hand side of the match arm. For example:
+
+```move
+public struct Wrapper(u64)
+
+fun add_under_wrapper_unless_equal(wrapper: Wrapper, x: u64): u64 {
+    match (wrapper) {
+        Wrapper(y) if (y == x) => Wrapper(y),
+        Wrapper(y) => y + x,
+    }
+}
+add_under_wrapper_unless_equal(Wrapper(1), 2); // returns Wrapper(3)
+add_under_wrapper_unless_equal(Wrapper(2), 3); // returns Wrapper(5)
+add_under_wrapper_unless_equal(Wrapper(3), 3); // returns Wrapper(3)
+```
+
+Patterns can be nested, and patterns can be or'd with other patterns. The `..` pattern is a special
+pattern that matches any number of fields in a struct or enum variant, but it can only occur within
+a constructor pattern, similarly the `mut` pattern can only be used within constructor patterns --
```

Comment:
Make `mut` a new sentence.

Reply:
I think I'll just remove the sentence since we don't really need to go into it here since there aren't any restrictions on it :)